[Blazor] Fixes issues with route precedence

javiercn opened the pull request:
#### Description

In 5.0 we introduced two features on **Blazor routing** that enable users to write routing templates that match paths with variable length segments. These two features are optional parameters `{parameter?}` and catch all parameters `{*catchall}`.

Our routing system ordered the routes based on precedence and the (now false) assumption that route templates would only match paths with an equal number of segments.

The implementation that we have worked for naïve scenarios but breaks on more real world scenarios. The change here includes fixes to the way we order the routes in the route table to match the expectations as well as fixes on the route matching algorithm to ensure we match routes with variable number of segments correctly.

#### Customer Impact

This was reported by customers on #27250

The impact is that a route with `{*catchall}` will prevent more specific routes like `/page/{parameter}` from being accessible.

There are no workarounds since precedence is a fundamental behavior of the routing system.

#### Regression?

No, these Blazor features were initially added in 5.0.

#### Risk

Low. These two features were just introduced in 5.0 and their usage is not as prevalent as in asp.net core routing. That said, it's important to fix them as otherwise we run the risk of diverting in behavior from asp.net core routing and Blazor routing, which is not something we want to do.

We have functional tests covering the area and we've added a significant amount of unit tests to validate the changes.

javiercn committed:
Fix routing precedence

javiercn committed:
Fix optional followed by catch-all

captainsafia commented on src/Components/Components/src/Routing/TemplateSegment.cs:
```diff
@@ -30,43 +29,55 @@ public TemplateSegment(string template, string segment, bool isParameter)
-                Constraints = tokens.Skip(1)
-                    .Select(token => RouteConstraint.Parse(template, segment, token))
-                    .ToArray();
+                Constraints = Array.Empty<RouteConstraint>();
```
Maybe set the default value for `Constraints` before we enter the if-statement here.

javiercn replied:
It's a bit weird, isn't it? since the other branch will have to set the constraints array to a different value.

captainsafia commented on src/Components/Components/test/Routing/RouteTableFactoryTests.cs:
```diff
@@ -676,26 +818,27 @@ public void SuppliesNullForUnusedHandlerParameters()
+                    Assert.Equal("{unrelated}", route.Template.TemplateText);
+                    Assert.Equal(Array.Empty<string>(), route.UnusedRouteParameterNames.OrderBy(id => id).ToArray());
                 });
+
```
It doesn't look like we have a test for validating how matching will work when the route table contains an catch-all parameter followed by an optional parameter. Let me know if I missed where this is written.

javiercn replied:
I believe we don't. That said, it's not correct to have an optional parameter followed by a catch-all, we just need a test that validate we throw when we parse the template.

captainsafia commented on src/Components/Components/src/Routing/TemplateSegment.cs:
```diff
@@ -30,43 +29,55 @@ public TemplateSegment(string template, string segment, bool isParameter)
 
-                Constraints = Array.Empty<RouteConstraint>();
+                    Value = tokens[0];
+                    IsOptional = tokens[^1].EndsWith("?");
```
I like the use of index ranges here!

captainsafia commented on src/Components/Components/src/Routing/RouteTableFactory.cs:
```diff
@@ -121,62 +120,84 @@ internal static int RouteComparison(RouteEntry x, RouteEntry y)
-                            return comparison;
-                        }
-                    }
+                if ((currentResult == -1 && HasRemainingOptionalSegments(yTemplate, minSegments + 1)) ||
```
Nit: Would be helpful from a readability perspective to set these booleans to variables, something like `shorterWithOptionals` etc.

javiercn replied:
hmm, I think it might be best to put an explaining comment rather than the explaining variable to explain why we are doing these checks, so I'll add that instead.

javiercn replied:
This got removed because length still matters, even for optional segments. So routes are only ambiguous when the only difference is the optionality of the parameters, which was already covered by this.

javiercn commented on src/Components/Components/src/Routing/RouteTableFactory.cs:
```diff
@@ -121,62 +120,84 @@ internal static int RouteComparison(RouteEntry x, RouteEntry y)
+                for (var i = firstRemainingSegmentIndex; i < template.Segments.Length; i++)
+                {
+                    var segment = template.Segments[i];
+                    if (!segment.IsOptional)
```
The method name might not be super clear, it's more like `AreAllRemainingSegmentsOptional` and by that it means literally `{parameter?}`

javiercn commented on src/Components/Components/src/Routing/RouteTableFactory.cs:
```diff
@@ -121,62 +120,84 @@ internal static int RouteComparison(RouteEntry x, RouteEntry y)
+            return xSegment switch
+            {
+                // Literal
+                { IsParameter: false, Constraints: { Length: 0 } } => 0,
```
Actually, this could be just `{IsParameter: false}`

SteveSandersonMS commented:
To build more confidence that the algorithmic changes don't have any unanticipated breaking effect on existing apps, I had a go at running the *old* tests against the *new* implementation. Five of the old tests fail with the new implementation, all of which may be intentional and for good reasons, but it's worth spelling out the exact details given this is going in a patch.

Here are the old tests that fail with the new algorithm:

 * `SuppliesNullForUnusedHandlerParameters`
    * My interpretation is that this fails because `RouteTable` used to sort by number of segments. Previously `/{unrelated}` would go before `/products/{id}` simply because `/{unrelated}` has fewer segments, even though `/products/{id}` could be regarded as more specific. In practice this never mattered because a one-segment template wouldn't match a two-segment URL. The new algorithm prefers specificity regardless of the number of segments and hence puts `/products/{id}` before `/{unrelated}`.
    * Conclusion: I think this change is fine because it doesn't seem like it can affect routing results

[WORK IN PROGRESS - I'M ADDING INFO HERE AS I FIGURE IT OUT]

javiercn committed:
Code review comments and more tests

javiercn committed:
Additional tests

SteveSandersonMS committed:
Fix typo in test

captainsafia commented:
@SteveSandersonMS Great summary, Steve! Thanks for taking the time to assemble that. I think the change in how we handle optional and non-optional constraints for the same parameter might end up being the most impactful breaking change for users. Since this is being put into a patch release, what's our approach for handling the breaking changes here?

SteveSandersonMS commented:
@captainsafia I think we'll need to make this change opt-in (and on by default in the project template for new apps).

SteveSandersonMS commented:
@javiercn I'm assuming I'm still finishing this off (and am working on it now), but if you preferred to take it back, please let me know! Just don't want us both to be working on it separately at the same time.

SteveSandersonMS committed:
Add "legacy" fork of the route matching logic taken from release/5.0

It's completely independent of the newer logic, because it's in a separate peer-level namespace. Nothing in the legacy namespace has a "using" for the newer namespace, and nothing in the newer namespace has a "using" for the legacy namespace. So they should not be able to reference each other accidentally.

SteveSandersonMS committed:
Add unit tests from release/5.0 for the legacy route matching code

SteveSandersonMS committed:
Update <Router> to use legacy routing by default, but opt into current routing via flag

SteveSandersonMS committed:
Unit test to show the router can switch between the two routing modes

SteveSandersonMS committed:
Use PreferExactMatches in all our E2E tests/samples

SteveSandersonMS committed:
Tweak XML doc format

SteveSandersonMS committed:
Use PreferExactMatches on project templates

SteveSandersonMS commented on src/Components/Components/src/Routing/Router.cs:
```diff
@@ -75,7 +76,20 @@ public class Router : IComponent, IHandleAfterRender, IDisposable
+        /// In .NET 6, this option will be removed and the router will always prefer exact matches.
+        /// </para>
+        /// </remarks>
+        [Parameter] public bool PreferExactMatches { get; set; }
```
After consideration, I think we should not name the new option after the package version number, because:

1. This is going in a patch, so the version would be something like `RoutingBehavior_5_0_1` which is ridiculously ugly to put in the default template
2. We can’t really use a name like `Current` because theoretically there could be an even newer incompatible version in the future

My proposal here of calling it `PreferExactMatches` has the benefits:

* It doesn't look particularly unpleasant in the default template. It sounds like a good thing, which it is.
* Even after we retire the option in 6.0 and remove it from intellisense, we can continue to allow people to pass  `PreferExactMatches="true"` (that just becomes a no-op, whereas `PreferExactMatches="false"` becomes an error). It doesn't make anyone's code look wrong to have this still in it, plus it doesn't imply the person *only* wants 5.0.1 behaviors and not newer ones
* If we ever need to fork the logic again, we can use the same pattern of naming the new option after its own behavior, and that's independent of this.

javiercn replied:
I would actually prefer that we only have a single flag for this.

I would hope we can avoid having to deal with a cross-section of things we can turn on/off and make it a multi value instead or something different that only allows you to set compatibility to a point in time.

SteveSandersonMS replied:
As long as we don't end up with more incompatible routing algorithm changes in 5.0.x, then there will only be a single flag. It's going away immediately in 6.0 - we're not carrying any complexity forwards into the long term with it.

javiercn replied:
Fair enough

SteveSandersonMS commented on src/Components/Components/src/Routing/LegacyRouteMatching/LegacyOptionalTypeRouteConstraint.cs:
```diff
@@ -1,20 +1,20 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-namespace Microsoft.AspNetCore.Components.Routing
+namespace Microsoft.AspNetCore.Components.LegacyRouteMatching
```
To minimize any risk that the two different route matching implementations interfere with one another, I've put them in separate peer-level namespaces, neither of which references the other.

I've also renamed all the legacy ones to `Legacy***` to avoid any risk that, in the future, somebody references one from the legacy namespace by mistake via the "Add using statement" codefix.

SteveSandersonMS replied:
Also, I know it's ugly having all this extra code. But as soon as this patch is merged to `master`, we can also do a follow-up change in `master` that removes everything from the `LegacyRouteMatching` directory and its corresponding tests, so we pretty much won't see this in all future work.

javiercn replied:
Sounds good to me, I would have done the same thing

javiercn reviewed (commented):
LGTM!

(I can't sign-off because I created the PR)

javiercn commented:
/AzurePipelines run AspNetCore-ci

mkArtakMSFT reviewed (approved):
Signing off on behalf of @javiercn 👍

SteveSandersonMS commented:
@mkArtakMSFT Looks like this is ready to merge.

Our test coverage for the project templates is very weak at the moment (seems that all the Blazor template tests are currently quarantined and failing), so I hope we have CTI coverage and will get preview builds of the 5.0.1 SDK to test manually before it ships.

mkArtakMSFT committed:
Merge branch 'release/5.0' into javiercn/blazor-routing
